src/commands/identity/listUserAssignedIdentities/FunctionAppUserAssignedIdentitiesListStep.ts

```diff
@@ -0,0 +1,109 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { type ManagedServiceIdentityClient } from '@azure/arm-msi';
+import { type ParsedSite } from '@microsoft/vscode-azext-azureappservice';
+import { createAuthorizationManagementClient, createManagedServiceIdentityClient, parseAzureResourceId, uiUtils, type ParsedAzureResourceId, type Role } from '@microsoft/vscode-azext-azureutils';
+import { AzureWizardPromptStep, nonNullProp, type IAzureQuickPickItem } from '@microsoft/vscode-azext-utils';
+import { ext } from '../../../extensionVariables';
+import { localize } from '../../../localize';
+import { type IFunctionAppUserAssignedIdentitiesContext } from './IFunctionAppUserAssignedIdentitiesContext';
+
+/**
+ * Wizard step to select a user-assigned managed identity from the parsed site of a function app.
+ * Upon selection, retrieves and stores the identity on the wizard context.
+ *
+ * @param role Optional. If provided, the function app will be pre-checked for an existing user assigned identity with this target role.
+ * If such an identity exists, it will be automatically assigned as a managed identity without prompting the user.
+ *
+ * @populates `context.managedIdentity`
+ */
+export class FunctionAppUserAssignedIdentitiesListStep<T extends IFunctionAppUserAssignedIdentitiesContext> extends AzureWizardPromptStep<T> {
+    private _msiClient: ManagedServiceIdentityClient;
+    private _hasTargetRole?: boolean;
+
+    constructor(readonly targetRole?: Role) {
+        super();
+    }
+
+    /**
+     * Indicates whether there is at least one user-assigned identity on the function app with the provided role.
+     * If no role is provided, or if the step has not yet been run, this will return `undefined`.
+     */
+    get hasIdentityWithTargetRole(): boolean | undefined {
+        return this._hasTargetRole;
+    }
+
+    // Verify if any of the existing user assigned identities for the function app have the required role already
+    public async configureBeforePrompt(context: T): Promise<void> {
+        if (!this.targetRole || !this.targetRole?.scopeId) {
+            this._hasTargetRole = undefined;
```

Comment:
Is there any functional difference between having this as undefined or false?

Reply:
Yeah, `undefined` after the step has been run implies no target role was provided to check for, whereas `false` implies a role was provided but not found.

That said this is a somewhat temporary solution, I have a better solution coming down the pipeline.  With this change in Azure Tools (https://github.com/microsoft/vscode-azuretools/pull/2036), once I incorporate the change, I no longer need to expose this getter because I can just run an async check in the role callback instead, I'll include that change in a near term PR.